misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJob.kt

```diff
@@ -2,23 +2,47 @@ package misk.jobqueue.sqs
 
 import com.amazonaws.services.sqs.model.Message
 import com.amazonaws.services.sqs.model.SendMessageRequest
+import com.squareup.moshi.Moshi
 import misk.jobqueue.Job
 import misk.jobqueue.QueueName
+import misk.moshi.adapter
 import misk.time.timed
 
 internal class SqsJob(
   override val queueName: QueueName,
   private val queues: QueueResolver,
   private val metrics: SqsMetrics,
+  private val moshi: Moshi,
   private val message: Message
 ) : Job {
   override val body: String = message.body
   override val id: String = message.messageId
-  override val attributes: Map<String, String> = message.messageAttributes.map { (key, value) ->
-    key to value.stringValue
-  }.toMap()
+  override val idempotenceKey: String by lazy {
+    jobqueueMetadata[JOBQUEUE_METADATA_IDEMPOTENCE_KEY]
+        ?: error("$JOBQUEUE_METADATA_IDEMPOTENCE_KEY not set in $JOBQUEUE_METADATA_ATTR")
+  }
+  override val attributes: Map<String, String> by lazy {
+    message.messageAttributes
+        .filter { (key, _) -> key != JOBQUEUE_METADATA_ATTR }
+        .map { (key, value) -> key to value.stringValue }.toMap()
+  }
 
   private val queue: ResolvedQueue = queues.getForReceiving(queueName)
+  private val jobqueueMetadata: Map<String, String> by lazy {
+    val metadata = message.messageAttributes[JOBQUEUE_METADATA_ATTR]
+    // NB: old/in-flight jobs enqueued prior to the introduction of _jobqueue-metadata won't have this property. In
+    // order to not change the contract later on (optional to non-optional), synthesize a reasonable replacement.
+    if (metadata == null) {
+      // TODO(bruno): drop fallback after rollout; error when metadata is not set.
+      mapOf(
+          JOBQUEUE_METADATA_ORIGIN_QUEUE to queueName.parentQueue.value,
+          // If the job had no metadata, there was no app-specified idempotence key;
+          // any sufficiently random value would work, so just use system-assigned id.
+          JOBQUEUE_METADATA_IDEMPOTENCE_KEY to id)
+    } else {
+      moshi.adapter<Map<String, String>>().fromJson(metadata.stringValue)!!
```

Comment:
I'd make the adapter a variable so we don't re-create it everytime

Reply:
Moshi [caches adapters](https://github.com/square/moshi/blob/master/moshi/src/main/java/com/squareup/moshi/Moshi.java#L66) (just to err on the safe side, double checked with `System.identityHashCode()` just to be sure, and multiple calls with the same arguments result in same object identity).